File: bindings/python/opentlv/src/opentlv/writer.py
"""Sequential writer over a self-managed, growable buffer."""

from __future__ import annotations

from typing import Union

import opentlv_native as _native

from opentlv.entry import Entry
from opentlv.error import BufferTooShortError, _from_native
from opentlv.fixed_format import FixedFormat
from opentlv.format import Format
from opentlv.tag import Tag

Value = Union[bytes, bytearray, memoryview]
AnyFormat = Union[Format, FixedFormat]

_INITIAL_CAPACITY = 64


class Writer:
    """A sequential writer that encodes TLV entries into a growable buffer.

    Unlike the C library's `tlv_writer_t`, which fills a caller-provided
    fixed-capacity buffer, `Writer` owns a `bytearray` that grows as needed:
    `write` never fails for lack of space, only for an error the format
    itself reports (for example an unsupported tag size).

    >>> writer = Writer()
    >>> writer.write(Tag(b"\\x01"), b"\\xaa\\xbb")
    >>> writer.write(Tag(b"\\x02"), b"")
    >>> writer.bytes()
    b'\\x01\\x02\\xaa\\xbb\\x02\\x00'
    """

    __slots__ = ("_buffer", "_pos", "_format")

    def __init__(self, format: AnyFormat = Format.DEFAULT) -> None:
        self._buffer = bytearray(_INITIAL_CAPACITY)
        self._pos = 0
        self._format = format

    @property
    def format(self) -> AnyFormat:
        """The wire format this writer encodes."""
        return self._format

    @property
    def position(self) -> int:
        """The number of bytes written so far."""
        return self._pos

    def _write_native(self, offset: int, tag_bytes: bytes, value: Value) -> int:
        if isinstance(self._format, FixedFormat):
            return _native.write_fixed(self._buffer, offset, tag_bytes, value,
                                       self._format.tag_size, self._format.length_size,
                                       self._format.big_endian)
        return _native.write(self._buffer, offset, tag_bytes, value, self._format)

    def write(self, tag: Union[Tag, bytes], value: Value) -> None:
        """Appends one entry with `tag` and `value`."""
        tag_bytes = tag.data if isinstance(tag, Tag) else tag
        try:
            written = self._write_native(self._pos, tag_bytes, value)
        except _native.Error as native_error:
            error = _from_native(native_error)
            if not (isinstance(error, BufferTooShortError) and error.required is not None):
                raise error from None
            self._grow(self._pos + error.required)
            try:
                written = self._write_native(self._pos, tag_bytes, value)
            except _native.Error as retry_error:
                raise _from_native(retry_error) from None
        self._pos += written

    def write_entry(self, entry: Entry) -> None:
        """Appends a decoded `Entry`, for example one produced by a `Reader`."""
        self.write(entry.tag, entry.value)

    def _grow(self, min_capacity: int) -> None:
        capacity = len(self._buffer)
        while capacity < min_capacity:
            capacity *= 2
        self._buffer.extend(bytes(capacity - len(self._buffer)))

    def bytes(self) -> bytes:
        """Returns the bytes written so far, as a new `bytes` object."""
        return bytes(self._buffer[:self._pos])

    def __len__(self) -> int:
        return self._pos

    def __repr__(self) -> str:
        return f"Writer(format={self._format!r}, position={self._pos})"


def encoded_size(tag: Union[Tag, bytes], value_length: int,
                 format: AnyFormat = Format.DEFAULT) -> int:
    """Returns the encoded size of an element with `tag` and a value of
    `value_length` bytes in `format`, without writing anything.

    >>> encoded_size(Tag(b"\\x01"), 3)
    5
    """
    tag_bytes = tag.data if isinstance(tag, Tag) else tag
    try:
        if isinstance(format, FixedFormat):
            return _native.encoded_size_fixed(tag_bytes, value_length, format.tag_size,
                                              format.length_size, format.big_endian)
        return _native.encoded_size(tag_bytes, value_length, format)
    except _native.Error as native_error:
        raise _from_native(native_error) from None

File: bindings/rust/opentlv/src/writer.rs
//! Sequential writer over a borrowed output buffer.

use std::marker::PhantomData;
use std::mem::MaybeUninit;
use std::slice;

use opentlv_native as native;

use crate::entry::Entry;
use crate::error::{Error, Result};
use crate::fixed_format::FixedFormat;
use crate::format::Format;
use crate::tag::Tag;

/// Returns the encoded size of an element with the given tag and value length
/// in `format`, without writing anything.
///
/// # Errors
///
/// [`Error::InvalidTagSize`] for an empty tag, [`Error::InvalidLength`] if the
/// total does not fit in `usize`, or any error of the format.
///
/// ```
/// use opentlv::{encoded_size, Format, Tag};
///
/// let tag = Tag::from_bytes(&[0x01]);
/// assert_eq!(encoded_size(&tag, 3, Format::Default).unwrap(), 5);
/// ```
pub fn encoded_size(tag: &Tag, value_len: usize, format: Format) -> Result<usize> {
    let mut size = 0usize;
    // SAFETY: `format.writer_raw()` points to a static format and `size` is
    // a valid, writable `usize`.
    let code =
        unsafe { native::tlv_encoded_size(tag.raw(), value_len, format.writer_raw(), &mut size) };
    Error::check(code)?;
    Ok(size)
}

/// Returns the encoded size of an element with the given tag and value length
/// in a [`FixedFormat`], without writing anything.
///
/// # Errors
///
/// Same as [`encoded_size`].
pub fn encoded_size_fixed(tag: &Tag, value_len: usize, format: &FixedFormat) -> Result<usize> {
    let mut size = 0usize;
    // SAFETY: `format.writer_raw()` points at storage owned by `format`,
    // borrowed for this call only, and `size` is a valid, writable `usize`.
    let code =
        unsafe { native::tlv_encoded_size(tag.raw(), value_len, format.writer_raw(), &mut size) };
    Error::check(code)?;
    Ok(size)
}

/// A sequential writer that encodes TLV entries into a byte slice.
///
/// The writer never allocates: it fills the caller's buffer and reports
/// [`Error::BufferTooShort`] when an entry does not fit. A failed write leaves
/// the position unchanged. Values are supplied as `&[u8]`.
///
/// ```
/// use opentlv::{Tag, Writer};
///
/// let mut buf = [0u8; 16];
/// let mut writer = Writer::new(&mut buf);
/// writer.write(&Tag::from_bytes(&[0x01]), &[0xAA, 0xBB]).unwrap();
/// writer.write(&Tag::from_bytes(&[0x02]), &[]).unwrap();
/// assert_eq!(writer.written(), &[0x01, 0x02, 0xAA, 0xBB, 0x02, 0x00]);
/// ```
#[derive(Debug)]
pub struct Writer<'a> {
    raw: native::tlv_writer_t,
    _buf: PhantomData<&'a mut [u8]>,
}

impl<'a> Writer<'a> {
    /// Creates a writer for the default format: a one-byte tag and a definite
    /// BER length.
    pub fn new(buf: &'a mut [u8]) -> Writer<'a> {
        Writer::with_format(buf, Format::Default)
    }

    /// Creates a writer for the given wire format.
    pub fn with_format(buf: &'a mut [u8], format: Format) -> Writer<'a> {
        let mut raw = MaybeUninit::<native::tlv_writer_t>::uninit();
        // SAFETY: `raw` is writable; `buf` is a valid slice (a non-null
        // pointer even when empty); `format.writer_raw()` points to a static
        // format.
        let code = unsafe {
            native::tlv_writer_init(
                raw.as_mut_ptr(),
                buf.as_mut_ptr(),
                buf.len(),
                format.writer_raw(),
            )
        };
        // Every argument is non-null and the built-in formats are complete, so
        // initialization cannot fail.
        assert_eq!(code, native::TLV_OK, "tlv_writer_init failed");
        Writer {
            // SAFETY: `tlv_writer_init` succeeded and initialized every field.
            raw: unsafe { raw.assume_init() },
            _buf: PhantomData,
        }
    }

    /// Creates a writer for a [`FixedFormat`]; `format` must outlive the writer.
    pub fn with_fixed_format(buf: &'a mut [u8], format: &'a FixedFormat) -> Writer<'a> {
        let mut raw = MaybeUninit::<native::tlv_writer_t>::uninit();
        // SAFETY: `raw` is writable; `buf` is a valid slice (a non-null
        // pointer even when empty); `format.writer_raw()` points at storage
        // owned by `format`, which the borrow checker keeps alive for `'a`.
        let code = unsafe {
            native::tlv_writer_init(
                raw.as_mut_ptr(),
                buf.as_mut_ptr(),
                buf.len(),
                format.writer_raw(),
            )
        };
        assert_eq!(code, native::TLV_OK, "tlv_writer_init failed");
        Writer {
            // SAFETY: `tlv_writer_init` succeeded and initialized every field.
            raw: unsafe { raw.assume_init() },
            _buf: PhantomData,
        }
    }

    /// Appends one entry with the given tag and value.
    ///
    /// # Errors
    ///
    /// [`Error::BufferTooShort`] if the entry does not fit in the remaining
    /// space, or any error of the format (for example an invalid tag). On error
    /// the position is unchanged, though bytes past it may have been modified.
    pub fn write(&mut self, tag: &Tag, value: &[u8]) -> Result<()> {
        // SAFETY: `self.raw` is initialized and its buffer is exclusively
        // borrowed for `'a`; `value` is a valid slice of `value.len()` bytes.
        let code = unsafe {
            native::tlv_writer_write(&mut self.raw, tag.raw(), value.as_ptr(), value.len())
        };
        Error::check(code)
    }

    /// Appends a decoded [`Entry`], for example one produced by a `Reader`.
    ///
    /// # Errors
    ///
    /// Same as [`Writer::write`].
    pub fn write_entry(&mut self, entry: &Entry<'_>) -> Result<()> {
        self.write(entry.tag(), entry.value())
    }

    /// Returns the number of bytes written so far.
    pub fn position(&self) -> usize {
        // SAFETY: `self.raw` was initialized by `tlv_writer_init`.
        unsafe { native::tlv_writer_size(&self.raw) }
    }

    /// Returns the total capacity of the output buffer in bytes.
    pub fn capacity(&self) -> usize {
        self.raw.capacity
    }

    /// Returns the number of bytes that can still be written.
    pub fn remaining(&self) -> usize {
        self.raw.capacity - self.position()
    }

    /// Returns the bytes written so far.
    pub fn written(&self) -> &[u8] {
        // SAFETY: the first `position()` bytes of the buffer were written by
        // the C library and are valid; the buffer is not mutated while `self`
        // is borrowed shared.
        unsafe { slice::from_raw_parts(self.raw.buf, self.position()) }
    }

    /// Ends writing and returns the written bytes with the buffer's lifetime.
    pub fn finish(self) -> &'a mut [u8] {
        // SAFETY: as in `written`; consuming `self` guarantees exclusive access
        // to the buffer, which was exclusively borrowed for `'a`.
        unsafe { slice::from_raw_parts_mut(self.raw.buf, self.position()) }
    }
}
